core/server/master/src/main/java/alluxio/underfs/MasterUfsManager.java

```diff
@@ -12,137 +12,65 @@
+import alluxio.underfs.UnderFileSystem.UfsMode;
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
-import java.util.Map.Entry;
-import java.util.concurrent.ConcurrentHashMap;
+import java.util.Set;
+import java.util.function.Supplier;
 
 import javax.annotation.concurrent.ThreadSafe;
 
 /**
  * A class that manages the UFS for master servers.
  */
 @ThreadSafe
-public final class MasterUfsManager extends AbstractUfsManager implements JournalEntryIterable {
+public final class MasterUfsManager extends AbstractUfsManager
+    implements JournalEntryIterable, JournalEntryReplayable {
   private static final Logger LOG = LoggerFactory.getLogger(MasterUfsManager.class);
 
-  /**
-   * {@link alluxio.underfs.UnderFileSystem.UfsMode} and mount ids corresponding to a physical ufs.
-   *
-   * Only the ufs modes part of this data structure is journaled.
-   */
-  public static class UfsState {
-    private UnderFileSystem.UfsMode mUfsMode;
-    private ConcurrentHashSet<Long> mMountIds;
-
-    /**
-     * Construct a new instance of UfsState w/ defaults.
-     */
-    UfsState() {
-      mUfsMode = UnderFileSystem.UfsMode.READ_WRITE;
-      mMountIds = new ConcurrentHashSet<>();
-    }
-
-    /**
-     * Add a mount id to list using physical ufs.
-     *
-     * @param mountId the mount id
-     */
-    void addMount(long mountId) {
-      mMountIds.addIfAbsent(mountId);
-    }
-
-    /**
-     * Remove mount id from list using physical ufs.
-     *
-     * @param mountId the mount id
-     * @return true, if mount list is empty
-     */
-    boolean removeMount(long mountId) {
-      mMountIds.remove(mountId);
-      return mMountIds.size() == 0;
-    }
+  private final State mState;
 
-    /**
-     * @return the physical ufs operation mode
-     */
-    UnderFileSystem.UfsMode getUfsMode() {
-      return mUfsMode;
-    }
+  /** A set of all managed ufs roots. */
+  private final Set<String> mUfsRoots;
 
-    /**
-     * Set the operation mode.
-     *
-     * @param ufsMode the ufs operation mode
-     */
-    void setUfsMode(UnderFileSystem.UfsMode ufsMode) {
-      mUfsMode = ufsMode;
-    }
-  }
-
-  // The physical ufs state for all managed mounts
-  private ConcurrentHashMap<String, UfsState> mPhysicalUfsToState =
-      new ConcurrentHashMap<>();
+  /** Mapping from mount ID to ufs root. */
+  private final Map<Long, String> mIdToRoot;
 
   /**
    * Constructs the instance of {@link MasterUfsManager}.
    */
-  public MasterUfsManager() {}
+  public MasterUfsManager() {
+    mState = new State();
+    mUfsRoots = new HashSet<>();
+    mIdToRoot = new HashMap<>();
+  }
 
   @Override
-  public void addMount(long mountId, final AlluxioURI ufsUri,
-                       final UnderFileSystemConfiguration ufsConf) {
+  public synchronized void addMount(long mountId, AlluxioURI ufsUri,
```

Comment:
what is the `synchronized` protecting?

Reply:
it protects all the state in the class. I'll added some `GuardedBy` annotations